airbyte-integrations/bases/base-singer/base_singer/source.py

```diff
@@ -22,15 +22,38 @@
 SOFTWARE.
 """
 
+import json
+from dataclasses import dataclass
 from typing import Dict, Generator, Type
 
 from airbyte_protocol import AirbyteCatalog, AirbyteConnectionStatus, AirbyteMessage, ConfiguredAirbyteCatalog, Status
-from base_python import AirbyteLogger, CatalogHelper, ConfigContainer, Source
+from base_python import AirbyteLogger, CatalogHelper, Source
 
 from .singer_helpers import Catalogs, SingerHelper, SyncModeInfo
 
 
+@dataclass
+class ConfigContainer:
+    rendered_config: object
+    rendered_config_path: str
+
+
 class SingerSource(Source):
+    # can be overridden to change an input file config
+    def transform_config(self, config_path: str, config: json):
+        return ConfigContainer(
+            rendered_config=config,
+            rendered_config_path=config_path,
+        )
+
+    # can be overridden to change an input catalog
+    def transform_catalog(self, catalog_path: str):
```

Comment:
what `transform_catalog` and `transform_state` needed for?

Reply:
They are overridden in singer to return path strings instead of actual catalogs and state objects